core/amber/src/main/scala/edu/uci/ics/amber/engine/common/promise/PromiseManager.scala

```diff
@@ -0,0 +1,236 @@
+          queuedInvocations.enqueue(payload)
+        }
+
+      // handle synchronous promise created by other promise
+      case PromiseInvocation(
+            ctx: ChildPromiseContext,
+            call: PromiseBody[_] with SynchronizedInvocation
+          ) =>
+        if (syncPromiseRoot == null || ctx.root == syncPromiseRoot) {
+          // if there is no other executing synchronous promise,
+          // or this promise has same root context (chain of promises),
+          // execute this one.
+          registerSyncPromise(ctx.root, ctx)
+          invokePromise(ctx, call)
+        } else {
+          // otherwise, enqueue it
+          queuedInvocations.enqueue(payload)
+        }
+
+      // trivial case
+      case p: PromiseInvocation =>
+        // set context
+        promiseContext = p.context
+        try {
+          // execute it
+          promiseHandler(p.call)
+        } catch {
+          case e: Throwable =>
+            // if error occurs, return it to the sender.
+            returning(e)
+        }
+    }
+
+    // execute queued sync promises
+    tryInvokeNextSyncPromise()
+  }
+
+  // send a control message to another actor, and keep the handle.
+  def schedule[T](cmd: PromiseBody[T], on: ActorVirtualIdentity): Promise[T] = {
+    val ctx = mkPromiseContext()
+    promiseID += 1
+    controlOutputPort.sendTo(on, PromiseInvocation(ctx, cmd))
+    val promise = WorkflowPromise[T](promiseContext)
+    unCompletedPromises(ctx) = promise
+    promise
+  }
+
+  // send a grouped control message to other actors, and keep the handle.
+  def schedule[T: ClassTag](seq: (PromiseBody[T], ActorVirtualIdentity)*): Promise[Seq[T]] = {
+    val promise = WorkflowPromise[Seq[T]](promiseContext)
+    if (seq.isEmpty) {
+      // if the sequence is empty, resolve the promise immediately
+      promise.setValue(Seq.empty)
+    } else {
+      unCompletedGroupPromises.add(
+        GroupedWorkflowPromise[T](promiseID, promiseID + seq.length, promise)
+      )
+      seq.foreach {
+        case (body, virtualIdentity) =>
+          val ctx = mkPromiseContext()
+          promiseID += 1
+          controlOutputPort.sendTo(virtualIdentity, PromiseInvocation(ctx, body))
+      }
+    }
+    promise
+  }
+
+  def returning(value: Any): Unit = {
+    // returning should be used at most once per context
+    if (promiseContext != null) {
+      controlOutputPort.sendTo(promiseContext.sender, ReturnPayload(promiseContext, value))
+      exitCurrentPromise()
+    }
+  }
+
+  def returning(): Unit = {
+    // returning should be used at most once per context
+    if (promiseContext != null) {
+      controlOutputPort.sendTo(
+        promiseContext.sender,
+        ReturnPayload(promiseContext, PromiseCompleted())
+      )
+      exitCurrentPromise()
+    }
+  }
+
+  def createLocalPromise[T](): WorkflowPromise[T] = {
+    val ctx = mkPromiseContext()
+    promiseID += 1
+    val promise = WorkflowPromise[T](promiseContext)
```

Comment:
This is confusing. Why are we creating a promiseContext `ctx` but the returned promise has some other context.

Reply:
Then I realized I cannot change line 139-141. We are creating call stacks. The current promise should know where to return by using `promiseContext`, and it schedules another call by creating a local context which is `ctx`, `ctx` is also sent to the destination so that call knows where to return. If that call returns, the current call may also return, so we will need `promiseContext` again.

Reply:
And I realized the pause logic is wrong because we used the wrong promise context. I pushed a fix for it.